netty/src/main/java/io/grpc/netty/WriteQueue.java

```diff
@@ -126,20 +119,20 @@ ChannelFuture enqueue(QueuedCommand command, ChannelPromise promise, boolean flu
    */
   private void flush() {
     try {
-      boolean flushed = false;
-      while (queue.drainTo(writeChunk, DEQUE_CHUNK_SIZE) > 0) {
-        while (writeChunk.size() > 0) {
-          QueuedCommand cmd = writeChunk.poll();
-          channel.write(cmd, cmd.promise());
+      QueuedCommand cmd;
+      int i = 0;
+      while ((cmd = queue.poll()) != null) {
+        channel.write(cmd, cmd.promise());
+        if (++i == DEQUE_CHUNK_SIZE) {
+          i = 0;
+          // Flush each chunk so we are releasing buffers periodically. In theory this loop
+          // might never end as new events are continuously added to the queue, if we never
+          // flushed in that case we would be guaranteed to OOM.
+          channel.flush();
         }
-        // Flush each chunk so we are releasing buffers periodically. In theory this loop
-        // might never end as new events are continuously added to the queue, if we never
-        // flushed in that case we would be guaranteed to OOM.
-        flushed = true;
-        channel.flush();
       }
-      if (!flushed) {
-        // Must flush at least once
+      // Must flush at least once
+      if (i != 0) {
```

Comment:
Some cases:

1. if there is nothing in the write queue, `i` will be 0 here and there will not be a flush.  I think this okay.
2. if `i` is a multiple of `DEQUE_CHUNK_SIZE`, then flush will be called in the loop above.
3. if `i` is not a multiple of `DEQUE_CHUNK_SIZE`, then flush may have been called if there were more then `DEQUE_CHUNK_SIZE` items taken from the queue.  If there were not, no flush would have been called.  Regardless, there have still be writes that need a flush.  This check ensures that case 2 didn't happen.